R/shared_constants.R

```diff
@@ -228,6 +228,68 @@ strip_names <- function(x) {
   x
 }
 
+extract_glued_symbols <- function(expr, interpret_glue) {
+  if (!isTRUE(interpret_glue)) {
+    return(character())
+  }
+  # TODO support more glue functions
+  # Package glue:
+  #  - glue_sql
+  #  - glue_safe
+  #  - glue_col
+  #  - glue_data
+  #  - glue_data_sql
+  #  - glue_data_safe
+  #  - glue_data_col
+  #
+  # Package stringr:
+  #  - str_interp
+  # NB: position() > 1 because position=1 is <expr><SYMBOL_FUNCTION_CALL>
+  glue_call_xpath <- "
+    descendant::SYMBOL_FUNCTION_CALL[text() = 'glue']
+      /parent::expr
+      /parent::expr[
+        not(SYMBOL_SUB[text() = '.envir' or text() = '.transform'])
+        and not(expr[position() > 1 and not(STR_CONST)])
+      ]
+  "
+  glue_calls <- xml_find_all(expr, glue_call_xpath)
+
+  glued_symbols <- new.env(parent = emptyenv())
+  for (glue_call in glue_calls) {
+    # TODO(michaelchirico): consider dropping tryCatch() here if we're more confident in our logic
+    parsed_call <- tryCatch(xml2lang(glue_call), error = unexpected_error, warning = unexpected_error)
+    parsed_call[[".envir"]] <- glued_symbols
+    parsed_call[[".transformer"]] <- symbol_extractor
+    # #1459: syntax errors in glue'd code are ignored with warning, rather than crashing lint
+    tryCatch(eval(parsed_call), error = parse_failure_warning)
+  }
+  names(glued_symbols)
+}
+
+unexpected_error <- function(cond) {
```

Comment:
These functions should be either local to `extract_glued_symbols()` or named more specific.

Reply:
Done. Pulled them out to avoid overhead of declaring several functions each time `extract_glued_symbols` is invoked.